src/mip/coefficient_strengthening.jl

```diff
@@ -0,0 +1,175 @@
+
+In case the constraint is ranged, i.e, cᵢ ⩽ aᵢ x ⩽ bᵢ, no coefficient strengthening is performed.
+
+The new coefficients are stored by updating ps.pb0 .
+If some coefficient is reduced to 0, it is kept as 0 in both ps.pb0.arows and ps.pb0.acols,
+and the number of nonzeros in each row/column is updated.
+
+If a row/column is reduced to 0 after this step, it will be set to be inactive.
+"""
+
+function maximal_activity(row::Row{T}, ucol::Vector{T}, lcol::Vector{T})::T where {T}
+    sup = zero(T)
+
+    for (j, aij) in zip(row.nzind, row.nzval)
+        if aij > zero(T)
+            sup += aij*ucol[j]
+        elseif aij < zero(T)
+            sup += aij*lcol[j]
+        end
+    end
+    return T(sup)
+end
+
+function minimal_activity(row::Row{T}, ucol::Vector{T}, lcol::Vector{T})::T where {T}
+    inf = zero(T)
+
+    for (j, aij) in zip(row.nzind, row.nzval)
+        if aij > zero(T)
+            inf += aij*lcol[j]
+        elseif aij < zero(T)
+            inf += aij*ucol[j]
+        end
+    end
+    return T(inf)
+end
+
+function upperbound_strengthening(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, max_act) where {T}
+    # perform coef strengthening for one constraints of the from a'x <= u
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    new_bound = ps.urow[i]
+    new_coef = a[j_index]
+    if a[j_index] > 0
+        d = new_bound - max_act - a[j_index]*(ps.ucol[j]-1)
+        if a[j_index] >= d > 0
+            new_coef = new_coef - d
+            new_bound -= d*ps.ucol[j]
+        end
+    elseif a[j_index] < 0
+        d = new_bound - max_act - a[j_index]*(ps.lcol[j]+1)
+        if -a[j_index] >= d > 0
+            new_coef = new_coef + d
+            new_bound += d*ps.lcol[j]
+        end
+    end
+    return new_coef, new_bound
+end
+
+function lowerbound_strengthening(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, min_act) where {T}
+    # perform coef strengthening for one constraints of the from l < = a'x
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    new_bound = ps.lrow[i]
+    new_coef = a[j_index]
+    if a[j_index] > 0
+        d = -new_bound + min_act + a[j_index]*(ps.lcol[j]+1)
+        if a[j_index] >= d > 0
+            new_coef = a[j_index] - d
+            new_bound -= d*ps.lcol[j]
+        end
+    elseif a[j_index] < 0
+        d = -new_bound + min_act + a[j_index]*(ps.ucol[j]-1)
+        if -a[j_index] >= d > 0
+            new_coef = a[j_index] + d
+            new_bound += d*ps.ucol[j]
+        end
+    end
+    return new_coef, new_bound
+end
+
+function zero_coefficient_strengthening!(ps::PresolveData{T}) where {T}
+    # perform coefficient stregthening but if there is a coefficient is reduced to 0
+    # it is still kept in the ps.pb0.arows and ps.pb0.acols
+
+    # keep track of index for each var fo update ps.acols
+    # use this to find which index of ps.pb0.acols[i].nzval to update
+    i_index = zeros(Int, ps.pb0.nvar)
+
+    for i in 1:ps.pb0.ncon
```

Comment:
Skip deactivated rows
```suggestion
    for i in 1:ps.pb0.ncon
        ps.rowflag[i] || continue
```

Reply:
All fixed, can you review it again ?

Reply:
Thanks! Looks good to me :)